third_party/zuho/zuho.js

```diff
@@ -15,20 +15,15 @@ const SHADERS = {
     uniform sampler2D uTex;
     varying vec2      vPos;
 
-    bool unproject(vec2, out vec3);
-
-    vec4 sample(float dx, float dy) {
-      vec2 p = vPos + uPxSize * vec2(dx, dy);
-      vec3 q;
-      if(unproject(p, q)) {
-        vec3 dir = normalize(uRot * q);
-        float u = (-0.5 / pi) * atan(dir[1], dir[0]) + 0.5;
-        float v = (1.0 / pi) * acos(dir[2]);
-        return texture2D(uTex, vec2(u, v));
-      }
-      else {
-        return vec4(0.0);
-      }
+    vec4 sample( float dx, float dy ) {
```

Comment:
yeah... I'll be honest: I only understand the whitespace changes in this file 😅 

Do we have any kind of inline comments to make this clearer?  Or tests to assert the correct behavior?

As is, I have no idea whether this has broken something

Reply:
hey, 
the zuho.js is a lightweight wrapper around a WebGL context, it is meant to display a collection of projections (formarly stored under zuho.Mapping) in the shader the `bool unproject(vec2, out vec3);` method was meant to be replaced with different implementations corresponding to the different projections ( `if( unproject(p, q) ){` in the next diff).
the goal of this unprojection was to convert the position of a 2D pixel into a 3D coordinate on a sphere to know where to sample the panorama texture.
with some projections, this would result in invalid coordinates, which in turn would give invalid texture sampling coordinate.
this condition was here to handle the invalid cases and fallback to a transparent pixel ( `}else{ return vec4(0.); }` in the next diff)
branching in a shader is usually considered bad practice but more importantly, the projection we use is always valid as you can see from the previous commit L.88 
so what we was "simply" to remove the test (`if(unproject(p,q){)else{}`) and inline the resulting 3D vector of the unproject method (` vec3 q = vec3(vPos + uPxSize * vec2(dx, dy), -1.0);` )
not sure if it's clearer but that's what happened here ^^'

Reply:
Thanks for the catch @nicoptere ! The only projection method we are using did not require this test and branching is costly, at all so it was not just dead code, but expensive dead code 😬. Which reminds me I need to start looking at @ jonlab's EAC projection implementation and figure out how to remove all the branching.

As for the whitespace changes in the shader, it's meant to save a few bytes in the JS package: because it's a string literal, it does not get obfuscated/minimzed by the JS compiler like the rest of the code. We could easily shred more bytes, but the shader would become even more obscure 😅

Reply:
So, are these added lines pulling in code that we'd previously trimmed from the zuho library?  Or is this net new code?

Reply:
There is no new logic in the shader with the latest implementation, only a small refactor/simplification. There were some heavier shader changes in the first proposed implementation (see [this commit](https://github.com/ampproject/amphtml/pull/30138/commits/fb243e10129a8ae6a0c70dfcb395c5bf348cf00e)) but now the new logic has moved entirely to the JS side with the `orientation` property and the new `euler_` method.

Note however that more substantial changes will be required in the shader when we add EAC projection mapping for the video support. And I have no idea how we can add tests to a webgl shader.

Reply:
🤔

Are there ways we can add a manual test at least?  Providing steps to test and expected results?

Reply:
Definitely something to think about when we make these changes. Maybe screenshots would work with Percy?

Reply:
Sounds good.  I can file an issue to follow up; LGTM for now.